src/tacacs_plus_fmt_plug.c

```diff
@@ -0,0 +1,337 @@
+} *cur_salt;
+
+static void init(struct fmt_main *self)
+{
+#ifdef _OPENMP
+	int omp_t = omp_get_max_threads();
+
+	self->params.min_keys_per_crypt *= omp_t;
+	omp_t *= OMP_SCALE;
+	self->params.max_keys_per_crypt *= omp_t;
+#endif
+	saved_key = mem_calloc(sizeof(*saved_key), self->params.max_keys_per_crypt);
+	any_cracked = 0;
+	cracked_size = sizeof(*cracked) * self->params.max_keys_per_crypt;
+	cracked = mem_calloc(cracked_size, 1);
+}
+
+static void done(void)
+{
+	MEM_FREE(saved_key);
+	MEM_FREE(cracked);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *ctcopy, *keeptr, *p;
+	int value, extra;
+
+	if (strncmp(ciphertext, FORMAT_TAG, TAG_LENGTH) != 0)
+		return 0;
+
+	ctcopy = strdup(ciphertext);
+	keeptr = ctcopy;
+
+	ctcopy += TAG_LENGTH;
+	if ((p = strtokm(ctcopy, "$")) == NULL) // version / type
+		goto err;
+	if (!isdec(p))
+		goto err;
+	value = atoi(p);
+	if (value != 0)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // pre_hash_data
+		goto err;
+	if (hexlenl(p, &extra) != 4 * 2 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // ciphertext
+		goto err;
+	if (hexlenl(p, &extra) > MAX_CIPHERTEXT_LENGTH * 2 || extra)
+		goto err;
+	if ((p = strtokm(NULL, "$")) == NULL)   // hash_data
+		goto err;
+	if (hexlenl(p, &extra) != 2 * 2 || extra)
+		goto err;
+
+
+	MEM_FREE(keeptr);
+	return 1;
+
+err:
+	MEM_FREE(keeptr);
+	return 0;
+}
+
+static void *get_salt(char *ciphertext)
+{
+	static struct custom_salt cs;
+	char *ctcopy = strdup(ciphertext);
+	char *keeptr = ctcopy;
+	char *p;
+	int i;
+
+	memset(&cs, 0, SALT_SIZE);
+	ctcopy += TAG_LENGTH;
+	p = strtokm(ctcopy, "$");
+	p = strtokm(NULL, "$");
+	cs.pre_hash_data_len = strlen(p) / 2;
+	for (i = 0; i < cs.pre_hash_data_len; i++)
+		cs.pre_hash_data[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "$");
+	cs.ctlen = strlen(p) / 2;
+	for (i = 0; i < cs.ctlen; i++)
+		cs.ciphertext[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	p = strtokm(NULL, "$");
+	cs.hash_data_len = 2;
+	for (i = 0; i < 2; i++)
+		cs.hash_data[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+	MD5_Init(&cs.pctx);
+	MD5_Update(&cs.pctx, cs.pre_hash_data, cs.pre_hash_data_len);
+
+	MEM_FREE(keeptr);
+
+	return &cs;
+}
+
+static int check_password(char *secret, struct custom_salt *cs)
+{
+	MD5_CTX cur;
+	unsigned char digest[MD5_DIGEST_LENGTH];
+	int i;
+	unsigned char status, flags;
+	unsigned short server_msg_len, data_len;
+	unsigned char cleartext[MD5_DIGEST_LENGTH];
+	int length;
+
+	length = (cs->ctlen < MD5_DIGEST_LENGTH ? cs->ctlen : MD5_DIGEST_LENGTH);
+	memcpy((void *)&cur, &cur_salt->pctx, sizeof(MD5_CTX));
+	MD5_Update(&cur, secret, strlen(secret));
+	MD5_Update(&cur, cs->hash_data, cs->hash_data_len);
+	MD5_Final(digest, &cur);
+
+	for (i = 0; i < length; i++) {
+		cleartext[i] = cs->ciphertext[i] ^ digest[i];
+	}
```

Comment:
Can easily speedup the XORs by using union's and XOR'ing two uint64_t's instead of 16 chars.

Reply:
Actually, only need to XOR one uint64_t for now, since we're not using anything beyond first 6 bytes for the check below. This will also eliminate the need to determine `length`.

Reply:
Does this look sane to you?

``` diff
diff --git a/src/tacacs_plus_fmt_plug.c b/src/tacacs_plus_fmt_plug.c
index 96f173a..f19fb2d 100644
--- a/src/tacacs_plus_fmt_plug.c
+++ b/src/tacacs_plus_fmt_plug.c
@@ -102,7 +102,10 @@ static struct custom_salt {
 	unsigned char pre_hash_data[8];
 	uint32_t hash_data_len;
 	unsigned char hash_data[8];
-	unsigned char ciphertext[MAX_CIPHERTEXT_LENGTH];
+	union {
+		uint64_t chunks[128];
+		unsigned char buf[MAX_CIPHERTEXT_LENGTH];
+	} ciphertext;
 } *cur_salt;
 
 static void init(struct fmt_main *self)
@@ -189,7 +192,7 @@ static void *get_salt(char *ciphertext)
 	p = strtokm(NULL, "$");
 	cs.ctlen = strlen(p) / 2;
 	for (i = 0; i < cs.ctlen; i++)
-		cs.ciphertext[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
+		cs.ciphertext.buf[i] = (atoi16[ARCH_INDEX(p[2 * i])] << 4) | atoi16[ARCH_INDEX(p[2 * i + 1])];
 	p = strtokm(NULL, "$");
 	cs.hash_data_len = 2;
 	for (i = 0; i < 2; i++)
@@ -205,27 +208,29 @@ static void *get_salt(char *ciphertext)
 static int check_password(int index, struct custom_salt *cs)
 {
 	MD5_CTX cur;
-	unsigned char digest[MD5_DIGEST_LENGTH];
-	int i;
+	union {
+		uint64_t chunks[MD5_DIGEST_LENGTH / sizeof(uint64_t)];
+		unsigned char buf[MD5_DIGEST_LENGTH];
+	} digest;
 	unsigned char status, flags;
 	unsigned short server_msg_len, data_len;
-	unsigned char cleartext[MD5_DIGEST_LENGTH];
-	int length;
+	union {
+		uint64_t chunks[MD5_DIGEST_LENGTH / sizeof(uint64_t)];
+		unsigned char buf[MD5_DIGEST_LENGTH];
+	} cleartext;
 
-	length = (cs->ctlen < MD5_DIGEST_LENGTH ? cs->ctlen : MD5_DIGEST_LENGTH);
 	memcpy((void *)&cur, &cur_salt->pctx, sizeof(MD5_CTX));
 	MD5_Update(&cur, saved_key[index], saved_len[index]);
 	MD5_Update(&cur, cs->hash_data, cs->hash_data_len);
-	MD5_Final(digest, &cur);
+	MD5_Final(digest.buf, &cur);
 
-	for (i = 0; i < length; i++) {
-		cleartext[i] = cs->ciphertext[i] ^ digest[i];
-	}
+	// XOR the first 8 bytes of "ciphertext" with "digest"
+	cleartext.chunks[0] = cs->ciphertext.chunks[0] ^ digest.chunks[0];
 
-	status = cleartext[0];
-	flags = cleartext[1];
-	server_msg_len = cleartext[3] | (cleartext[2] << 8);
-	data_len = cleartext[5] | (cleartext[4] << 8);
+	status = cleartext.buf[0];
+	flags = cleartext.buf[1];
+	server_msg_len = cleartext.buf[3] | (cleartext.buf[2] << 8);
+	data_len = cleartext.buf[5] | (cleartext.buf[4] << 8);
 	if(((status >= 0x01 && status <= 0x07) || status == 0x21) && (flags == 0x01 || flags == 0x00) &&
 		(6 + server_msg_len + data_len == cs->ctlen)) {
 		return 1;
```

Reply:
Looks sane, thanks. Instead of `uint64_t chunks[128];` please use `uint64_t chunks[(MAX_CIPHERTEXT_LENGTH + 7) / 8];` or alternatively replace those `chunks` arrays with simple `uint64_t chunk0;` fields in the union as we don't use further elements yet. We don't have to match the char array's size - we just need to avoid exceeding it by more than a fraction of the last uint64_t's size. If you keep the arrays, you could also call the union fields `u64` and `c` instead of `chunks` and `buf`. Oh, and I don't see why you use separate `digest` and `cleartext` variables - you could instead do `digest.u64[0] ^= ...` and use that in the following code. And if you do keep both variables, then don't repeat effectively the same `union { ...` twice, but instead list the two variables comma-separated on one line.